src/base/UGraphic.pas

```diff
@@ -554,43 +557,79 @@ procedure InitializeScreen;
   end; // case
 
   // If there is a resolution in Parameters, use it, else use the Ini value
-  I := Params.Resolution;
-  if (I <> -1) then
-    S := IResolution[I]
+  // check for a custom resolution (in the format of WIDTHxHEIGHT) or try validating ID from TIni
+  if ParseResolutionString(Params.CustomResolution, W, H) then
+    Log.LogStatus(Format('Use custom resolution from Command line: %d x %d', [W, H]), 'SDL_SetVideoMode')
+  else if Ini.GetResolution(Params.Resolution, S) and ParseResolutionString(S, W, H) then
+    Log.LogStatus(Format('Use resolution by index from command line: %d x %d [%d]', [W, H, Params.Resolution]), 'SDL_SetVideoMode')
   else
-    S := IResolution[Ini.Resolution];
+  begin
+    Log.LogStatus('Use config resolution', 'SDL_SetVideoMode');
+    S := Ini.GetResolution(W, H);
+  end;
 
-  I := Pos('x', S);
-  W := StrToInt(Copy(S, 1, I-1));
-  H := StrToInt(Copy(S, I+1, 1000));
   if ((Screens > 1) and not Split) then
   	W := W * Screens;
 
-  Log.LogStatus('SDL_SetVideoMode', 'Initialize3D');
+  Log.LogStatus('Creating window', 'SDL_SetVideoMode');
 
   // check whether to start in fullscreen or windowed mode.
   // The command-line parameters take precedence over the ini settings.
   Fullscreen := ((Ini.FullScreen = 1) or (Params.ScreenMode = scmFullscreen)) and
                 not (Params.ScreenMode = scmWindowed);
 
+  // TODO: use SDL renderer (for proper scale in "real fullscreen"). Able to choose rendering mode (OpenGL, OpenGL ES, Direct3D)
+  // TODO: Use real fullscreen mode with custom resolution, use native desktop fullscreen mode as "Windowed fullscreen" / borderless mode
   if Fullscreen then
   begin
-    Log.LogStatus('SDL_SetVideoMode', 'Set Video Mode...   Full Screen');
+    Log.LogStatus('Set Video Mode...   Full Screen', 'SDL_SetVideoMode');
     screen := SDL_CreateWindow('UltraStar Deluxe loading...',
-           SDL_WINDOWPOS_CENTERED,SDL_WINDOWPOS_CENTERED, W, H, SDL_WINDOW_OPENGL or SDL_WINDOW_FULLSCREEN_DESKTOP or SDL_WINDOW_RESIZABLE);
+              SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 0, 0, SDL_WINDOW_OPENGL or SDL_WINDOW_FULLSCREEN_DESKTOP);
```

Comment:
> Otherwise the window will not be resizable after starting in fullscreen and then switching to windowed mode on most operating systems.

Do you know how to test this case properly? (which OS)

I think instead of applying the resizable flag for the fullscreen window (which isn't even a recognizable window), the flag and/or resolution could be changed whenever the users switches or system/window event occurs.